dbcon/joblist/batchprimitiveprocessor-jl.cpp

```diff
@@ -703,6 +703,12 @@ bool BatchPrimitiveProcessorJL::countThisMsg(messageqcpp::ByteStream& in) const
     const uint8_t* data = in.buf();
     uint32_t offset = sizeof(ISMPacketHeader) + sizeof(PrimitiveHeader); // skip the headers
 
+    ISMPacketHeader* hdr = (ISMPacketHeader*)(data);
+
+    // Exit early if PrimProc has thrown an exception
+    if (hdr->Status > 0)
+        return true;
```

Comment:
This is a very good change but you should also log a message explaining that PP throws an exception running a primitive.

Reply:
This check is added because the below assertion in this function `idbassert(in.length() > offset);` was failing when the buffer contained an error message. We need to skip the rest of this function execution in case of an error.

The error logging is handled by the calling function TupleBPS::receiveMultiPrimitiveMessages. Specifically this code block:

```
if (bs->length() == 0 || hdr->Status > 0)
                {
                    /* PM errors mean this should abort right away instead of draining the PM backlog */
                    tplLock.lock();

                    if (bs->length() == 0)
                    {
                        errorMessage(IDBErrorInfo::instance()->errorMsg(ERR_PRIMPROC_DOWN));
                        status(ERR_PRIMPROC_DOWN);
                    }
                    else
                    {
                        string errMsg;

                        bs->advance(sizeof(ISMPacketHeader) + sizeof(PrimitiveHeader));
                        *bs >> errMsg;
                        status(hdr->Status);
                        errorMessage(errMsg);
                    }

                    abort_nolock();
                    goto out;
                }
```

Reply:
If PP has thrown an exception there will be no error message from PP and it looks like an assertion in ExeMgr w/o any note about PP. That is why I asked about an explicit message.

Reply:
This is not true. PP sends the error info (code and message) to ExeMgr in case an exception is raised by PP. Refer to the catch blocks in `BatchPrimitiveProcessor::execute()`. These blocks call `writeErrorMsg` function which log the error info to bytestream, which are then read by the ExeMgr on the receiving side.

Reply:
I suggest you to try to throw an exception from BatchPrimitiveProcessor::execute() and see how the logging output helps you to understand that PP throws.

Reply:
I believe that was a good moment to finally start logging something usefull in case of PP throw. I believe I should do it on my own.